modelscope_agent/llm/dashscope_llm.py

```diff
@@ -29,38 +30,60 @@ def generate(self,
-                    prompt=llm_artifacts,
-                    stream=False,
-                    **self.generate_cfg)
-                llm_result = CustomOutputWrapper.handle_message_text_completion(
-                    response)
-            return llm_result
-        except NotImplementedError as e:
-            error = traceback.format_exc()
-            error_msg = f'LLM error with input {llm_artifacts} \n dashscope error: {str(e)} with traceback {error}'
-            print(error_msg)
-            raise RuntimeError(error)
+        error_message_list = []
+        for i in range(3):
+            print('call generate at {} time'.format(i + 1))
+            try:
+                if self.agent_type == AgentType.Messages:
+                    messages = llm_artifacts if len(
+                        llm_artifacts) > 0 else DEFAULT_MESSAGE
+                    self.generate_cfg['use_raw_prompt'] = False
+                    response = dashscope.Generation.call(
+                        model=self.model,
+                        messages=messages,
+                        # set the random seed, optional, default to 1234 if not set
+                        seed=random.randint(1, 10000),
+                        result_format=
+                        'message',  # set the result to be "message" format.
+                        stream=False,
+                        **self.generate_cfg)
+                    if response.status_code == HTTPStatus.OK:
+                        llm_result = CustomOutputWrapper.handle_message_chat_completion(
+                            response)
+                        return llm_result
+                    else:
+                        err_msg = 'Error Request id: %s, Code: %d, status: %s, message: %s' % (
+                            response.request_id, response.status_code,
+                            response.code, response.message)
+                        print(err_msg)
+                        error_message_list.append(err_msg)
+                        time.sleep(i * 2 + 1)
+                else:
+                    response = Generation.call(
+                        model=self.model,
+                        prompt=llm_artifacts,
+                        stream=False,
+                        **self.generate_cfg)
+                    if response.status_code == HTTPStatus.OK:
+                        llm_result = CustomOutputWrapper.handle_message_text_completion(
+                            response)
+                        return llm_result
+                    else:
+                        err_msg = 'Error Request id: %s, Code: %d, status: %s, message: %s' % (
+                            response.request_id, response.status_code,
+                            response.code, response.message)
+                        print(err_msg)
+                        error_message_list.append(err_msg)
+                        time.sleep(i * 2 + 1)
+            except Exception as e:
```

Comment:
it is better to catch certain error and retry,  while some error should not, such as if a user forget to config dashscope api key, there is no need to retry

Reply:
you are right, I think 4xx status code may need not retry.